src/main/java/com/hivemq/client/internal/mqtt/handler/publish/incoming/MqttSubscriptionFlowTree.java

```diff
@@ -48,52 +43,73 @@
     public void subscribe(
             final @NotNull MqttTopicFilterImpl topicFilter, final @Nullable MqttSubscribedPublishFlow flow) {
 
-        final MqttTopicLevel level = MqttTopicLevel.root(topicFilter);
         final TopicTreeEntry entry = (flow == null) ? null : new TopicTreeEntry(flow, topicFilter);
-        if (rootNode == null) {
-            rootNode = new TopicTreeNode(ROOT_LEVEL, level, entry);
-        } else {
-            rootNode.subscribe(level, entry);
+        final MqttTopicIterator topicIterator = MqttTopicIterator.of(topicFilter);
+        TopicTreeNode node = rootNode;
+        if (node == null) {
+            rootNode = node = new TopicTreeNode(null, null);
+        }
+        while (node != null) {
+            node = node.subscribe(topicIterator, entry);
         }
     }
 
     @Override
     public void remove(final @NotNull MqttTopicFilterImpl topicFilter, final @Nullable MqttSubscribedPublishFlow flow) {
-        if ((rootNode != null) && rootNode.remove(MqttTopicLevel.root(topicFilter), flow)) {
-            rootNode = null;
+        final MqttTopicIterator topicIterator = MqttTopicIterator.of(topicFilter);
+        TopicTreeNode node = rootNode;
+        while (node != null) {
+            node = node.remove(topicIterator, flow);
         }
+        compact();
     }
 
     @Override
     public void unsubscribe(
             final @NotNull MqttTopicFilterImpl topicFilter,
             final @Nullable Consumer<MqttSubscribedPublishFlow> unsubscribedCallback) {
```

Comment:
Here you use your old formatting.
Maybe decide for one and reformat every method declaration.

In IntelliJ this can be defined under `Preferences - Code Style - Java - Method declaration parameters`. 👍

Reply:
Ah, I see now that if the parameters are too long you move them to a new line.
This seems fine 👍